sstables/trie.cc

```diff
@@ -0,0 +1,787 @@
+    auto bpp = bits_per_pointer_arr[type];
+    auto n_children = uint8_t(raw[1]);
+    auto idx = std::lower_bound(&raw[2], &raw[2 + n_children], transition) - &raw[2];
+    if (idx < n_children) {
+        return {idx, raw[2 + idx], read_offset(raw.subspan(2+n_children), idx, bpp)};
+    } else {
+        return {idx, std::byte(0), 0};
+    }
+}
+
+// We want to be always inlined so that bits_per_pointer is substituted with a constant,
+// and the read_offset can be simplified.
+[[gnu::always_inline]]
+static lookup_result find_child_dense(int type, const_bytes raw, std::byte transition) {
+    auto start = int(raw[1]);
+    auto idx = std::max<int>(0, int(transition) - start);
+    auto dense_span = uint64_t(raw[2]) + 1;
+    auto bpp = bits_per_pointer_arr[type];
+    while (idx < int(dense_span)) {
+        if (auto off = read_offset(raw.subspan(3), idx, bpp)) {
+            return {idx, std::byte(start + idx), off};
+        } else {
+            ++idx;
+        }
+    }
+    return {dense_span, std::byte(0), 0};
+}
+
+// Looks up the first child of this node greater with transition greater or equal to the given one.
+// If such child doesn't exist, the `idx` of the result will be negative.
+static lookup_result find_child(const_bytes raw, std::byte transition) {
+    auto type = uint8_t(raw[0]) >> 4;
+    switch (type) {
+    case PAYLOAD_ONLY:
+        return {0, std::byte(0), 0};
+    case SINGLE_NOPAYLOAD_4:
+        if (transition <= raw[1]) {
+            return {0, raw[1], uint64_t(raw[0]) & 0xf};
+        }
+        return {1, std::byte(0), 0};
+    case SINGLE_8:
+        if (transition <= raw[1]) {
+            return {0, raw[1], uint64_t(raw[2])};
+        }
+        return {1, std::byte(0), 0};
+    case SINGLE_NOPAYLOAD_12:
+        if (transition <= raw[2]) {
+            return {0, raw[2], (uint64_t(raw[0]) & 0xf) << 8 | uint64_t(raw[1])};
+        }
+        return {1, std::byte(0), 0};
+    case SINGLE_16:
+        if (transition <= raw[1]) {
+            return {0, raw[1], uint64_t(raw[1]) << 8 | uint64_t(raw[2])};
+        }
+        return {1, std::byte(0), 0};
+    case SPARSE_8:
+        return find_child_sparse(type, raw, transition);
+    case SPARSE_12:
+        return find_child_sparse(type, raw, transition);
+    case SPARSE_16:
+        return find_child_sparse(type, raw, transition);
+    case SPARSE_24:
+        return find_child_sparse(type, raw, transition);
+    case SPARSE_40:
+        return find_child_sparse(type, raw, transition);
+    case DENSE_12:
+        return find_child_dense(type, raw, transition);
+    case DENSE_16:
+        return find_child_dense(type, raw, transition);
+    case DENSE_24:
+        return find_child_dense(type, raw, transition);
+    case DENSE_32:
+        return find_child_dense(type, raw, transition);
+    case DENSE_40:
+        return find_child_dense(type, raw, transition);
+    case LONG_DENSE:
+        return find_child_dense(type, raw, transition);
```

Comment:
Perhaps move the bpp lookup here, so the callees can be a single function, tail-called, reducing cache footprint.